Issue with TClass object managed in case of multi-threaded 'rapid' set of dlopen/dlclose on same library.

As reported at https://github.com/root-project/roottest/pull/946#issuecomment-1519645135 and can be seen in the test
`roottest-root-meta-race-12552-exec_dlopenrace-auto`

 The relevant output is below and the problem is a (real) quirk of the test. It linked to the repeated unload and loading of the same library in parallel. During the unloading and loading cycle the TClass object representing the user classes (from that library) are created and deleted and it seems that the pointer value is captured and used outside the main lock (and this is done concurrently we the deletion). This needs to be fixed but is unlikely to happen in the wild.

```
==1544443==ERROR: AddressSanitizer: heap-use-after-free on address 0x616000617c20 at pc 0x7f8857c557ae bp 0x7f88265b3e60 sp 0x7f88265b3e50
READ of size 4 at 0x616000617c20 thread T1
    #0 0x7f8857c557ad in TClass::Size() const /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:5706:8
    #1 0x7f8851923388 in TBufferFile::ReadFastArray(void*, TClass const*, int, TMemberStreamer*, TClass const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TBufferFile.cxx:1605:25
    #2 0x7f8851cc28c1 in TStreamerInfoActions::ReadSTLObjectWiseFastArray(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*, short, unsigned int) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TStreamerInfoActions.cxx:1364:11
    #3 0x7f8851cc28c1 in int TStreamerInfoActions::ReadSTL<&(TStreamerInfoActions::ReadSTLMemberWiseSameClass(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*, short)), &(TStreamerInfoActions::ReadSTLObjectWiseFastArray(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*, short, unsigned int))>(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TStreamerInfoActions.cxx:1407:10
    #4 0x7f8851931ee7 in TStreamerInfoActions::TConfiguredAction::operator()(TBuffer&, void*) const /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/inc/TStreamerInfoActions.h:123:17
    #5 0x7f8851931ee7 in TBufferFile::ApplySequence(TStreamerInfoActions::TActionSequence const&, void*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TBufferFile.cxx:3580:10
    #6 0x7f8851930ca7 in TBufferFile::ReadClassBuffer(TClass const*, void*, TClass const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TBufferFile.cxx:3499:4
    #7 0x7f8851929ff7 in TClass::Streamer(void*, TBuffer&, TClass const*) const /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/inc/TClass.h:610:7
    #8 0x7f8851929ff7 in TBufferFile::ReadObjectAny(TClass const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TBufferFile.cxx:2473:14
    #9 0x7f8857bc179d in TObjArray::Streamer(TBuffer&) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/cont/src/TObjArray.cxx:468:29
    #10 0x7f8851b7c3bd in TClass::Streamer(void*, TBuffer&, TClass const*) const /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/inc/TClass.h:610:7
    #11 0x7f8851b7c3bd in TKey::ReadObjectAny(TClass const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TKey.cxx:1108:11
    #12 0x7f8851a61884 in TDirectoryFile::GetObjectChecked(char const*, TClass const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TDirectoryFile.cxx:1111:25
    #13 0x7f88400d8d51 in void TDirectory::GetObject<TObjArray>(char const*, TObjArray*&) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/inc/TDirectory.h:214:18
    #14 0x7f88400d8d51 in TCling::LoadPCMImpl(TFile&) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:1769:12
    #15 0x7f88400dbc28 in TCling::LoadPCM(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:1874:4
    #16 0x7f88400e2139 in TCling::RegisterModule(char const*, char const**, char const**, char const*, char const*, void (*)(), std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > const&, char const**, bool, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:2299:7
    #17 0x7f8857aa5a33 in TROOT::RegisterModule(char const*, char const**, char const**, char const*, char const*, void (*)(), std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > const&, char const**, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/src/TROOT.cxx:2557:15
    #18 0x7f88253a69de in (anonymous namespace)::TriggerDictionaryInitialization_libUser_Impl() /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/roottest/root/meta/race-12552/libUser.cxx:187:7
    #19 0x7f88253a537a in (anonymous namespace)::DictInit::DictInit() /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/roottest/root/meta/race-12552/libUser.cxx:195:7
    #20 0x7f88253a537a in __cxx_global_var_init.2 /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/roottest/root/meta/race-12552/libUser.cxx:197:5
    #21 0x7f88253a537a in _GLOBAL__sub_I_libUser.cxx /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/roottest/root/meta/race-12552/libUser.cxx
    #22 0x7f88584a3059 in call_init.part.0 /usr/src/debug/glibc-2.28-220.el8.x86_64/elf/dl-init.c:72:3
    #23 0x7f88584a3159 in call_init /usr/src/debug/glibc-2.28-220.el8.x86_64/elf/dl-init.c:118:11
    #24 0x7f88584a3159 in _dl_init /usr/src/debug/glibc-2.28-220.el8.x86_64/elf/dl-init.c:119:5
    #25 0x7f88565ac9bb in _dl_catch_exception (/lib64/libc.so.6+0x1689bb) (BuildId: f92f9985bf742e6e1ea71e6851f27783568e23d4)
    #26 0x7f88584aa88d in dl_open_worker /usr/src/debug/glibc-2.28-220.el8.x86_64/elf/dl-open.c:794:5
    #27 0x7f88584aa88d in dl_open_worker /usr/src/debug/glibc-2.28-220.el8.x86_64/elf/dl-open.c:757:1
    #28 0x7f88565ac963 in _dl_catch_exception (/lib64/libc.so.6+0x168963) (BuildId: f92f9985bf742e6e1ea71e6851f27783568e23d4)
    #29 0x7f88584aaa70 in _dl_open /usr/src/debug/glibc-2.28-220.el8.x86_64/elf/dl-open.c:876:17
    #30 0x7f8856a21f89 in dlopen_doit (/lib64/libdl.so.2+0xf89) (BuildId: a330b0475c03c7e7e095d17fbca122d945012730)
    #31 0x7f88565ac963 in _dl_catch_exception (/lib64/libc.so.6+0x168963) (BuildId: f92f9985bf742e6e1ea71e6851f27783568e23d4)
    #32 0x7f88565aca22 in _dl_catch_error (/lib64/libc.so.6+0x168a22) (BuildId: f92f9985bf742e6e1ea71e6851f27783568e23d4)
    #33 0x7f8856a2252d in _dlerror_run (/lib64/libdl.so.2+0x152d) (BuildId: a330b0475c03c7e7e095d17fbca122d945012730)
    #34 0x7f8856a22029 in dlopen@@GLIBC_2.2.5 (/lib64/libdl.so.2+0x1029) (BuildId: a330b0475c03c7e7e095d17fbca122d945012730)
    #35 0x496e15 in __interceptor_dlopen.part.267 asan_interceptors.cpp.o
    #36 0x7f8826a590c2  (<unknown module>)
    #37 0x7f8826a59b29  (<unknown module>)
    #38 0x7f8826a59abe  (<unknown module>)
    #39 0x7f8826a599fd  (<unknown module>)
    #40 0x7f8826a598df  (<unknown module>)
    #41 0x7f8826a598c3  (<unknown module>)
    #42 0x7f8857491b12  (/lib64/libstdc++.so.6+0xc2b12) (BuildId: 67383d84bc84f9c3d978093e405689f3fef20eca)
    #43 0x7f8856e351c9 in start_thread (/lib64/libpthread.so.0+0x81c9) (BuildId: faf67e21cdc2fe4c5de2be3d1ab54935cf9361f0)
    #44 0x7f885647de72 in clone (/lib64/libc.so.6+0x39e72) (BuildId: f92f9985bf742e6e1ea71e6851f27783568e23d4)

0x616000617c20 is located 416 bytes inside of 560-byte region [0x616000617a80,0x616000617cb0)
freed by thread T1 here:
    #0 0x511c48 in operator delete(void*) (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/bin/root.exe+0x511c48) (BuildId: 38590f4d5d08c122d0e0d8161ece8eeec84b2e5e)
    #1 0x7f8857a43f8f in TObject::operator delete(void*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/src/TObject.cxx:1078:7
    #2 0x7f8857c60240 in TClass::Init(char const*, short, std::type_info const*, TVirtualIsAProxy*, char const*, char const*, int, int, ClassInfo_t*, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx
    #3 0x7f8857c631e3 in TClass::TClass(char const*, short, std::type_info const&, TVirtualIsAProxy*, char const*, char const*, int, int, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:1329:4
    #4 0x7f8857c82f15 in ROOT::CreateClass(char const*, short, std::type_info const&, TVirtualIsAProxy*, char const*, char const*, int, int) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:5877:15
    #5 0x7f8857cb19f9 in ROOT::Internal::TDefaultInitBehavior::CreateClass(char const*, short, std::type_info const&, TVirtualIsAProxy*, char const*, char const*, int, int) const /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/inc/Rtypes.h:179:17
    #6 0x7f8857caf2cc in ROOT::TGenericClassInfo::GetClass() /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TGenericClassInfo.cxx:250:31
    #7 0x7f88400e26dd in TCling::ProcessClassesToUpdate() /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:1994:24
    #8 0x7f88400e26dd in TCling::RegisterModule(char const*, char const**, char const**, char const*, char const*, void (*)(), std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > const&, char const**, bool, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:2339:4
    #9 0x7f8857aa5a33 in TROOT::RegisterModule(char const*, char const**, char const**, char const*, char const*, void (*)(), std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > const&, char const**, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/src/TROOT.cxx:2557:15
    #10 0x7f8823177c7c in (anonymous namespace)::TriggerDictionaryInitialization_libHist_Impl() /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/hist/hist/G__Hist.cxx:15989:7
    #11 0x7f8822d01f29 in (anonymous namespace)::DictInit::DictInit() /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/hist/hist/G__Hist.cxx:15997:7
    #12 0x7f8822d01f29 in __cxx_global_var_init.284 /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/hist/hist/G__Hist.cxx:15999:5
    #13 0x7f8822d01f29 in _GLOBAL__sub_I_G__Hist.cxx /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/hist/hist/G__Hist.cxx
    #14 0x7f88584a3059 in call_init.part.0 /usr/src/debug/glibc-2.28-220.el8.x86_64/elf/dl-init.c:72:3

previously allocated by thread T1 here:
    #0 0x511128 in operator new(unsigned long) (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/bin/root.exe+0x511128) (BuildId: 38590f4d5d08c122d0e0d8161ece8eeec84b2e5e)
    #1 0x7f8857aacabe in TStorage::ObjectAlloc(unsigned long) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/src/TStorage.cxx:292:19
    #2 0x7f8857c82ee9 in TObject::operator new(unsigned long) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/inc/TObject.h:181:46
    #3 0x7f8857c82ee9 in ROOT::CreateClass(char const*, short, std::type_info const&, TVirtualIsAProxy*, char const*, char const*, int, int) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:5877:11
    #4 0x7f8857cb19f9 in ROOT::Internal::TDefaultInitBehavior::CreateClass(char const*, short, std::type_info const&, TVirtualIsAProxy*, char const*, char const*, int, int) const /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/inc/Rtypes.h:179:17
    #5 0x7f8857caf2cc in ROOT::TGenericClassInfo::GetClass() /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TGenericClassInfo.cxx:250:31
    #6 0x7f8857c6f398 in TClass::GetClass(char const*, bool, bool, unsigned long, unsigned long) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:3039:26
    #7 0x7f8857c594a0 in TClass::GetClass(char const*, bool, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:2970:11
    #8 0x7f8857c594a0 in TBuildRealData::Inspect(TClass*, char const*, char const*, void const*, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:850:28
    #9 0x7f88400ec4fa in TCling::InspectMembers(TMemberInspector&, void const*, TClass const*, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:2909:12
    #10 0x7f8857c67884 in TClass::CallShowMembers(void const*, TMemberInspector&, bool) const /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:2225:24
    #11 0x7f8857c5a795 in TClass::BuildRealData(void*, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:2080:11
    #12 0x7f88519308bb in TBufferFile::ReadClassBuffer(TClass const*, void*, TClass const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TBufferFile.cxx:3467:45
    #13 0x7f8851929ff7 in TClass::Streamer(void*, TBuffer&, TClass const*) const /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/inc/TClass.h:610:7
    #14 0x7f8851929ff7 in TBufferFile::ReadObjectAny(TClass const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TBufferFile.cxx:2473:14
    #15 0x7f8857bc179d in TObjArray::Streamer(TBuffer&) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/cont/src/TObjArray.cxx:468:29
    #16 0x7f8851b7c3bd in TClass::Streamer(void*, TBuffer&, TClass const*) const /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/inc/TClass.h:610:7
    #17 0x7f8851b7c3bd in TKey::ReadObjectAny(TClass const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TKey.cxx:1108:11
    #18 0x7f8851a61884 in TDirectoryFile::GetObjectChecked(char const*, TClass const*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/io/io/src/TDirectoryFile.cxx:1111:25
    #19 0x7f88400d8d51 in void TDirectory::GetObject<TObjArray>(char const*, TObjArray*&) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/inc/TDirectory.h:214:18
    #20 0x7f88400d8d51 in TCling::LoadPCMImpl(TFile&) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:1769:12
    #21 0x7f88400dbc28 in TCling::LoadPCM(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:1874:4
    #22 0x7f88400e2139 in TCling::RegisterModule(char const*, char const**, char const**, char const*, char const*, void (*)(), std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > const&, char const**, bool, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:2299:7
    #23 0x7f8857aa5a33 in TROOT::RegisterModule(char const*, char const**, char const**, char const*, char const*, void (*)(), std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > const&, char const**, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/src/TROOT.cxx:2557:15
    #24 0x7f8823d849de in GCC_except_table867 G__Tree.cxx
    #25 0x7f8823d8337a in GCC_except_table867 G__Tree.cxx
    #26 0x7f88584a3059 in call_init.part.0 /usr/src/debug/glibc-2.28-220.el8.x86_64/elf/dl-init.c:72:3

Thread T1 created by T0 here:
    #0 0x43c752 in __interceptor_pthread_create (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/bin/root.exe+0x43c752) (BuildId: 38590f4d5d08c122d0e0d8161ece8eeec84b2e5e)
    #1 0x7f8857491df8 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/lib64/libstdc++.so.6+0xc2df8) (BuildId: 67383d84bc84f9c3d978093e405689f3fef20eca)
    #2 0x7f8826a593ff  (<unknown module>)
    #3 0x7f8826a59340  (<unknown module>)
    #4 0x7f8826a5c028  (<unknown module>)
    #5 0x7f8840597a97 in cling::IncrementalExecutor::executeWrapper(llvm::StringRef, cling::Value*) const (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/lib/libCling.so+0x36cfa97) (BuildId: 31284afedbe992b553f91ae4294ea7d1470eea60)
    #6 0x7f8840413a92 in cling::Interpreter::RunFunction(clang::FunctionDecl const*, cling::Value*) (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/lib/libCling.so+0x354ba92) (BuildId: 31284afedbe992b553f91ae4294ea7d1470eea60)
    #7 0x7f884040d746 in cling::Interpreter::EvaluateInternal(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cling::CompilationOptions, cling::Value*, cling::Transaction**, unsigned long) (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/lib/libCling.so+0x3545746) (BuildId: 31284afedbe992b553f91ae4294ea7d1470eea60)
    #8 0x7f88406c8e0f in cling::MetaSema::actOnxCommand(llvm::StringRef, llvm::StringRef, cling::Value*) (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/lib/libCling.so+0x3800e0f) (BuildId: 31284afedbe992b553f91ae4294ea7d1470eea60)
    #9 0x7f88406f4f7b in cling::MetaParser::isXCommand(cling::MetaSema::ActionResult&, cling::Value*) (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/lib/libCling.so+0x382cf7b) (BuildId: 31284afedbe992b553f91ae4294ea7d1470eea60)
    #10 0x7f88406f3fd4 in cling::MetaParser::isCommand(cling::MetaSema::ActionResult&, cling::Value*) (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/lib/libCling.so+0x382bfd4) (BuildId: 31284afedbe992b553f91ae4294ea7d1470eea60)
    #11 0x7f88406bf72e in cling::MetaProcessor::process(llvm::StringRef, cling::Interpreter::CompilationResult&, cling::Value*, bool) (/home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/lib/libCling.so+0x37f772e) (BuildId: 31284afedbe992b553f91ae4294ea7d1470eea60)
    #12 0x7f88400e9b7e in HandleInterpreterException(cling::MetaProcessor*, char const*, cling::Interpreter::CompilationResult&, cling::Value*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:2428:29
    #13 0x7f88400e878a in TCling::ProcessLine(char const*, TInterpreter::EErrorCode*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:2572:22
    #14 0x7f88400ef37a in TCling::ProcessLineSynch(char const*, TInterpreter::EErrorCode*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/metacling/src/TCling.cxx:3518:17
    #15 0x7f88579cd523 in TApplication::ExecuteFile(char const*, int*, bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/src/TApplication.cxx:1661:30
    #16 0x7f88579cb264 in TApplication::ProcessLine(char const*, bool, int*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/base/src/TApplication.cxx:1506:14
    #17 0x7f885825c26b in TRint::ProcessLineNr(char const*, char const*, int*) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/rint/src/TRint.cxx:820:11
    #18 0x7f885825b68d in TRint::Run(bool) /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/rint/src/TRint.cxx:461:22
    #19 0x514b4a in main /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/main/src/rmain.cxx:84:12
    #20 0x7f885647ed84 in __libc_start_main (/lib64/libc.so.6+0x3ad84) (BuildId: f92f9985bf742e6e1ea71e6851f27783568e23d4)

SUMMARY: AddressSanitizer: heap-use-after-free /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/root/core/meta/src/TClass.cxx:5706:8 in TClass::Size() const
Shadow bytes around the buggy address:
  0x0c2c800baf30: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x0c2c800baf40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c800baf50: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c800baf60: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c800baf70: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c2c800baf80: fd fd fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c800baf90: fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa fa
  0x0c2c800bafa0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c800bafb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c800bafc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c800bafd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1544443==ABORTING

-- END TEST OUTPUT --
CMake Error at /home/sftnight/build/night/LABEL/ROOT-centos8/SPEC/asan/V/master/build/RootTestDriver.cmake:186 (message):
  got exit code 1 but expected 0


Standard Output

-- TEST COMMAND --

--- a/core/base/inc/Rtypes.h
+++ b/core/base/inc/Rtypes.h
@@ -105,6 +105,7 @@ template <class Tmpl> TBuffer &operator<<(TBuffer &buf, const Tmpl *obj);
 namespace ROOT {
 
    class TGenericClassInfo;
+   class TClassAlt;
    typedef void *(*NewFunc_t)(void *);
    typedef void *(*NewArrFunc_t)(Long_t size, void *arena);
    typedef void  (*DelFunc_t)(void *);
@@ -124,7 +125,8 @@ namespace ROOT {
                         DictFuncPtr_t dict, Int_t pragmabits);
    extern void RemoveClass(const char *cname, TClass *cl);
    extern void ResetClassVersion(TClass*, const char*, Short_t);
-   extern void AddClassAlternate(const char *normName, const char *alternate);
+   extern ROOT::TClassAlt* AddClassAlternate(const char *normName, const char *alternate);
+   extern void RemoveClassAlternate(ROOT::TClassAlt*);
 
    extern TNamed *RegisterClassTemplate(const char *name,
                                         const char *file, Int_t line);

--- a/core/base/inc/Rtypes.h
+++ b/core/base/inc/Rtypes.h
@@ -122,7 +122,7 @@ namespace ROOT {
                               Int_t dl, Int_t il);
    extern void AddClass(const char *cname, Version_t id, const std::type_info &info,
                         DictFuncPtr_t dict, Int_t pragmabits);
-   extern void RemoveClass(const char *cname);
+   extern void RemoveClass(const char *cname, TClass *cl);
    extern void ResetClassVersion(TClass*, const char*, Short_t);
    extern void AddClassAlternate(const char *normName, const char *alternate);
 
@@ -153,7 +153,7 @@ namespace ROOT {
       virtual void Register(const char *cname, Version_t id,
                             const std::type_info &info,
                             DictFuncPtr_t dict, Int_t pragmabits) const = 0;
-      virtual void Unregister(const char *classname) const = 0;
+      virtual void Unregister(const char *classname, TClass *cl) const = 0;
       virtual TClass *CreateClass(const char *cname, Version_t id,
                                   const std::type_info &info, TVirtualIsAProxy *isa,
                                   const char *dfil, const char *ifil,
@@ -168,8 +168,8 @@ namespace ROOT {
          ROOT::AddClass(cname, id, info, dict, pragmabits);
       }
 
-      void Unregister(const char *classname) const override {
-         ROOT::RemoveClass(classname);
+      void Unregister(const char *classname, TClass *cl) const override {
+         ROOT::RemoveClass(classname, cl);
       }
 
       TClass *CreateClass(const char *cname, Version_t id,

--- a/core/clingutils/src/TClingUtils.cxx
+++ b/core/clingutils/src/TClingUtils.cxx
@@ -1998,15 +1998,15 @@ void ROOT::TMetaUtils::WriteClassInit(std::ostream& finalString,
    /////////////////////////////////////////////////////////////////////////////
 
    if (cl.GetRequestedName()[0] && classname != cl.GetRequestedName()) {
-      finalString << "\n" << "      ::ROOT::AddClassAlternate(\""
-                  << classname << "\",\"" << cl.GetRequestedName() << "\");\n";
+      finalString << "\n" << "      instance.AdoptAlternate(::ROOT::AddClassAlternate(\""
+                  << classname << "\",\"" << cl.GetRequestedName() << "\"));\n";
    }
 
    if (!cl.GetDemangledTypeInfo().empty()
          && cl.GetDemangledTypeInfo() != classname
          && cl.GetDemangledTypeInfo() != cl.GetRequestedName()) {
-      finalString << "\n" << "      ::ROOT::AddClassAlternate(\""
-                  << classname << "\",\"" << cl.GetDemangledTypeInfo() << "\");\n";
+      finalString << "\n" << "      instance.AdoptAlternate(::ROOT::AddClassAlternate(\""
+                  << classname << "\",\"" << cl.GetDemangledTypeInfo() << "\"));\n";
 
    }
 

--- a/core/cont/inc/TClassTable.h
+++ b/core/cont/inc/TClassTable.h
@@ -68,27 +68,28 @@ public:
 
    ~TClassTable();
 
-   static void          Add(const char *cname, Version_t id,
-                            const std::type_info &info, DictFuncPtr_t dict,
-                            Int_t pragmabits);
-   static void          Add(TProtoClass *protoClass);
-   static void          AddAlternate(const char *normname, const char *alternate);
-   static char         *At(UInt_t index);
-   int                  Classes();
-   static Bool_t        Check(const char *cname, std::string &normname);
-   static Version_t     GetID(const char *cname);
-   static Int_t         GetPragmaBits(const char *name);
-   static DictFuncPtr_t GetDict(const char *cname);
-   static DictFuncPtr_t GetDict(const std::type_info& info);
-   static DictFuncPtr_t GetDictNorm(const char *cname);
-   static TProtoClass  *GetProto(const char *cname);
-   static TProtoClass  *GetProtoNorm(const char *cname);
-   static void          Init();
-   static char         *Next();
-   void                 Print(Option_t *option="") const override;
-   static void          PrintTable();
-   static void          Remove(const char *cname);
-   static void          Terminate();
+   static void             Add(const char *cname, Version_t id,
+                               const std::type_info &info, DictFuncPtr_t dict,
+                               Int_t pragmabits);
+   static void             Add(TProtoClass *protoClass);
+   static ROOT::TClassAlt *AddAlternate(const char *normname, const char *alternate);
+   static char            *At(UInt_t index);
+   int                     Classes();
+   static Bool_t           Check(const char *cname, std::string &normname);
+   static Version_t        GetID(const char *cname);
+   static Int_t            GetPragmaBits(const char *name);
+   static DictFuncPtr_t    GetDict(const char *cname);
+   static DictFuncPtr_t    GetDict(const std::type_info& info);
+   static DictFuncPtr_t    GetDictNorm(const char *cname);
+   static TProtoClass     *GetProto(const char *cname);
+   static TProtoClass     *GetProtoNorm(const char *cname);
+   static void             Init();
+   static char            *Next();
+   void                    Print(Option_t *option="") const override;
+   static void             PrintTable();
+   static void             Remove(const char *cname);
+   static void             RemoveAlternate(ROOT::TClassAlt *alt);
+   static void             Terminate();
 
    ClassDefOverride(TClassTable,0)  //Table of known classes
 };

--- a/core/cont/inc/TClassTable.h
+++ b/core/cont/inc/TClassTable.h
@@ -98,7 +98,7 @@ R__EXTERN TClassTable *gClassTable;
 namespace ROOT {
    extern void AddClass(const char *cname, Version_t id, DictFuncPtr_t dict,
                         Int_t pragmabits);
-   extern void RemoveClass(const char *cname);
+   extern void RemoveClass(const char *cname, TClass *cl);
 }
 
 #endif

--- a/core/cont/src/TClassTable.cxx
+++ b/core/cont/src/TClassTable.cxx
@@ -521,8 +521,8 @@ void TClassTable::AddAlternate(const char *normName, const char *alternate)
             fprintf(stderr,"Error in TClassTable::AddAlternate: "
                     "Second registration of %s with a different normalized name (old: '%s', new: '%s')\n",
                     alternate, a->fNormName, normName);
-            return;
          }
+         return;
       }
    }
 

--- a/core/cont/src/TClassTable.cxx
+++ b/core/cont/src/TClassTable.cxx
@@ -504,7 +504,7 @@ void TClassTable::Add(TProtoClass *proto)
 
 ////////////////////////////////////////////////////////////////////////////////
 
-void TClassTable::AddAlternate(const char *normName, const char *alternate)
+ROOT::TClassAlt* TClassTable::AddAlternate(const char *normName, const char *alternate)
 {
    // This will be set at the lastest during TROOT construction, so before
    // any threading could happen.
@@ -522,11 +522,40 @@ void TClassTable::AddAlternate(const char *normName, const char *alternate)
                     "Second registration of %s with a different normalized name (old: '%s', new: '%s')\n",
                     alternate, a->fNormName, normName);
          }
-         return;
+         return nullptr;
       }
    }
 
    fgAlternate[slot] = new TClassAlt(alternate,normName,fgAlternate[slot]);
+   return fgAlternate[slot];
+}
+
+////////////////////////////////////////////////////////////////////////////////
+///
+void TClassTable::RemoveAlternate(ROOT::TClassAlt *alt)
+{
+   if (!alt || !gClassTable)
+      return;
+
+   std::lock_guard<std::mutex> lock(GetClassTableMutex());
+
+   UInt_t slot = ROOT::ClassTableHash(alt->fName, fgSize);
+
+   if (!fgAlternate[slot])
+      return;
+
+   if (fgAlternate[slot] == alt)
+      fgAlternate[slot] = alt->fNext.release();
+   else {
+      for (TClassAlt *a = fgAlternate[slot]; a; a = a->fNext.get()) {
+         if (a->fNext.get() == alt) {
+            a->fNext.swap( alt->fNext );
+	    assert( alt == alt->fNext.get());
+	    alt->fNext.release();
+	 }
+      }
+   }
+   delete alt;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
@@ -886,13 +915,24 @@ void ROOT::AddClass(const char *cname, Version_t id,
 /// Global function called by GenerateInitInstance.
 /// (see the ClassImp macro).
 
-void ROOT::AddClassAlternate(const char *normName, const char *alternate)
+ROOT::TClassAlt* ROOT::AddClassAlternate(const char *normName, const char *alternate)
 {
    if (!TROOT::Initialized() && !gClassTable) {
       GetDelayedAddClassAlternate().emplace_back(normName, alternate);
+      // If a library is loaded before gROOT is initialized we can assume
+      // it is hard linked along side libCore (or is libCore) thus can't
+      // really be unloaded.
+      return nullptr;
    } else {
-      TClassTable::AddAlternate(normName, alternate);
+      return TClassTable::AddAlternate(normName, alternate);
    }
+}
+
+void ROOT::RemoveClassAlternate(TClassAlt *alt)
+{
+   // This routine is meant to be called (indirectly) by dlclose so we
+   // we are guaranteed that the library initialization has completed.
+   TClassTable::RemoveAlternate(alt);
 }
 
 ////////////////////////////////////////////////////////////////////////////////

--- a/core/cont/src/TClassTable.cxx
+++ b/core/cont/src/TClassTable.cxx
@@ -975,13 +975,17 @@ TNamed *ROOT::RegisterClassTemplate(const char *name, const char *file,
 
    TString classname(name);
    Ssiz_t loc = classname.Index("<");
-   if (loc >= 1) classname.Remove(loc);
+   if (loc >= 1)
+      classname.Remove(loc);
+   TNamed *reg = (TNamed*)table.FindObject(classname);
    if (file) {
-      TNamed *obj = new TNamed((const char*)classname, file);
-      obj->SetUniqueID(line);
-      table.Add(obj);
-      return obj;
+      if (reg)
+         reg->SetTitle(file);
+      else {
+         reg = new TNamed((const char*)classname, file);
+         table.Add(reg);
+      }
+      reg->SetUniqueID(line);
    }
-
-   return (TNamed*)table.FindObject(classname);
+   return reg;
 }

--- a/core/cont/src/TClassTable.cxx
+++ b/core/cont/src/TClassTable.cxx
@@ -942,7 +942,7 @@ void ROOT::ResetClassVersion(TClass *cl, const char *cname, Short_t newid)
 /// Global function called by the dtor of a class's init class
 /// (see the ClassImp macro).
 
-void ROOT::RemoveClass(const char *cname)
+void ROOT::RemoveClass(const char *cname, TClass *oldcl)
 {
    // don't delete class information since it is needed by the I/O system
    // to write the StreamerInfo to file
@@ -952,14 +952,8 @@ void ROOT::RemoveClass(const char *cname)
       // pointer is now invalid ....
       // We still keep the TClass object around because TFile needs to
       // get to the TStreamerInfo.
-      if (gROOT && gROOT->GetListOfClasses()) {
-         TObject *pcname;
-         if ((pcname = gROOT->GetListOfClasses()->FindObject(cname))) {
-            TClass *cl = dynamic_cast<TClass*>(pcname);
-            if (cl)
-               cl->SetUnloaded();
-         }
-      }
+      if (oldcl)
+         oldcl->SetUnloaded();
       TClassTable::Remove(cname);
    }
 }

--- a/core/meta/inc/TGenericClassInfo.h
+++ b/core/meta/inc/TGenericClassInfo.h
@@ -71,6 +71,7 @@ namespace ROOT {
       Detail::TCollectionProxyInfo *fCollectionStreamerInfo;
       std::vector<ROOT::Internal::TSchemaHelper>  fReadRules;
       std::vector<ROOT::Internal::TSchemaHelper>  fReadRawRules;
+      std::vector<ROOT::TClassAlt*>               fAlternate;
 
    public:
       TGenericClassInfo(const char *fullClassname,
@@ -116,6 +117,7 @@ namespace ROOT {
 
       TClass                           *IsA(const void *obj);
 
+      void                              AdoptAlternate(ROOT::TClassAlt *alt);
       Short_t                           AdoptStreamer(TClassStreamer*);
       Short_t                           AdoptCollectionProxy(TVirtualCollectionProxy*);
       void                              AdoptCollectionProxyInfo(Detail::TCollectionProxyInfo*);

--- a/core/meta/src/TGenericClassInfo.cxx
+++ b/core/meta/src/TGenericClassInfo.cxx
@@ -218,7 +218,7 @@ namespace Internal {
       fIsA = nullptr;
       using ROOT::Internal::gROOTLocal;
       if (!gROOTLocal || !gROOTLocal->Initialized() || !gROOTLocal->GetListOfClasses()) return;
-      if (fAction) GetAction().Unregister(GetClassName());
+      if (fAction) GetAction().Unregister(GetClassName(), fClass);
    }
 
    const Internal::TInitBehavior &TGenericClassInfo::GetAction() const

--- a/core/meta/src/TGenericClassInfo.cxx
+++ b/core/meta/src/TGenericClassInfo.cxx
@@ -214,11 +214,14 @@ namespace Internal {
       delete fCollectionProxyInfo;
       delete fCollectionStreamerInfo;
       delete fStreamer;
-      if (!fClass) delete fIsA; // fIsA is adopted by the class if any.
+      if (!fClass)
+         delete fIsA; // fIsA is adopted by the class if any.
       fIsA = nullptr;
       using ROOT::Internal::gROOTLocal;
-      if (!gROOTLocal || !gROOTLocal->Initialized() || !gROOTLocal->GetListOfClasses()) return;
-      if (fAction) GetAction().Unregister(GetClassName(), fClass);
+      if (!gROOTLocal || !gROOTLocal->Initialized() || !gROOTLocal->GetListOfClasses())
+         return;
+      if (fAction)
+         GetAction().Unregister(GetClassName(), fClass);
    }
 
    const Internal::TInitBehavior &TGenericClassInfo::GetAction() const

--- a/core/meta/src/TGenericClassInfo.cxx
+++ b/core/meta/src/TGenericClassInfo.cxx
@@ -220,6 +220,8 @@ namespace Internal {
       using ROOT::Internal::gROOTLocal;
       if (!gROOTLocal || !gROOTLocal->Initialized() || !gROOTLocal->GetListOfClasses())
          return;
+      for(auto alt : fAlternate)
+         ROOT::RemoveClassAlternate(alt);
       if (fAction)
          GetAction().Unregister(GetClassName(), fClass);
    }
@@ -422,6 +424,11 @@ namespace Internal {
       ROOT::ResetClassVersion(fClass, GetClassName(),version);
       fVersion = version;
       return version;
+   }
+
+   void TGenericClassInfo::AdoptAlternate(ROOT::TClassAlt *alt)
+   {
+      fAlternate.push_back(alt);
    }
 
    void TGenericClassInfo::AdoptCollectionProxyInfo(TCollectionProxyInfo *info)

--- a/core/metacling/src/TCling.cxx
+++ b/core/metacling/src/TCling.cxx
@@ -1851,7 +1851,13 @@ void TCling::LoadPCM(std::string pcmFileNameFullPath)
 
       cling::Interpreter::PushTransactionRAII deserRAII(GetInterpreterImpl());
       LoadPCMImpl(pcmMemFile);
-      fPendingRdicts.erase(pendingRdict);
+      // Currently the module file are never unloaded (even if the library is
+      // unloaded) and, of course, never reloaded.
+      // Consequently, we must NOT remove the `pendingRdict` from the list
+      // of pending dictionary, otherwise if a library is unloaded and then
+      // reload we will be unable to update properly the TClass object
+      // (because we wont be able to load the rootpcm file by executing the
+      // above lines)
 
       return;
    }

